File: C/Searching/Searching_04.c
/*
===============================================================================
Program Name    : Bi-Directional Search
Program Number  : 04
File Name       : Searching_04.c
Description     : This program implements Bi-Directional Search to search for
                  a particular element by comparing elements from both ends
                  of a dynamically allocated array.

Date            : August 2026
Language        : C
Author          : Ishwari Vijaykumar Surve
===============================================================================

Concepts Used:
1. Structure
2. Pointer to Structure
3. Typedef
4. Dynamic Memory Allocation
5. Bi-Directional Search
6. Array Traversal

Key Learning Points:
- Understand the concept of Bi-Directional Search.
- Search elements from both ends of the array.
- Compare the starting and ending elements simultaneously.
- Analyze the time and space complexity of the searching technique.

===============================================================================
*/

#include<stdio.h>
#include<stdlib.h>

struct Searching
{
    int *Arr;
    int iSize;
};

typedef struct Searching SEARCHING;
typedef struct Searching *PSEARCHING;

void Accept(PSEARCHING first)
{
    int i = 0;

    printf("Enter the Elements:\n");

    for(i = 0; i < first->iSize; i++)
    {
        scanf("%d",&first->Arr[i]);
    }
}

void Display(PSEARCHING first)
{
    int i = 0;

    printf("Elements of the Array are:\n");

    for(i = 0; i < first->iSize; i++)
    {
        printf("%d\n",first->Arr[i]);
    }
}

int BiDirectionalSearch(PSEARCHING first, int iNo)
{
    int iStart = 0;
    int iEnd = first->iSize - 1;

    while(iStart <= iEnd)
    {
        if((first->Arr[iStart] == iNo) ||
           (first->Arr[iEnd] == iNo))
        {
            return 1;
        }

        iStart++;
        iEnd--;
    }

    return 0;
}

int main()
{
    SEARCHING sobj;
    int iRet = 0;

    sobj.iSize = 5;

    sobj.Arr = (int *)malloc(sizeof(int) * sobj.iSize);

    Accept(&sobj);
    Display(&sobj);

    iRet = BiDirectionalSearch(&sobj,30);

    if(iRet == 1)
    {
        printf("Element is Present\n");
    }
    else
    {
        printf("There is no such element\n");
    }

    free(sobj.Arr);

    return 0;
}

/*
===============================================================================
Expected Output:
===============================================================================

Enter the Elements:
10
20
30
40
50

Elements of the Array are:
10
20
30
40
50

Element is Present

===============================================================================
Memory Representation:
===============================================================================

              sobj
        +------------------+
        | Arr ------------ |--------+
        | iSize = 5        |        |
        +------------------+        |
                                    v
                         +------+------+------+------+------+
                         |  10  |  20  |  30  |  40  |  50  |
                         +------+------+------+------+------+
                           Arr[0] Arr[1] Arr[2] Arr[3] Arr[4]
                             ^                         ^
                           iStart                    iEnd

Step 1:
Compare Arr[0] and Arr[4]
10 != 30
50 != 30

Step 2:
Compare Arr[1] and Arr[3]
20 != 30
40 != 30

Step 3:
Compare Arr[2] and Arr[2]
30 == 30

Element Found

===============================================================================
Algorithm:
===============================================================================

1. Start the program.
2. Define the Searching structure.
3. Create a SEARCHING object.
4. Set the array size.
5. Allocate memory dynamically for the array.
6. Accept the array elements from the user.
7. Display the array elements.
8. Initialize iStart to 0.
9. Initialize iEnd to iSize - 1.
10. Compare the elements at iStart and iEnd with the search element.
11. If either element matches, return 1.
12. Increment iStart.
13. Decrement iEnd.
14. Repeat the process while iStart is less than or equal to iEnd.
15. If no match is found, return 0.
16. Display whether the element is present or absent.
17. Release the allocated memory using free().
18. Stop the program.

===============================================================================
Program Flow:
===============================================================================

        Start
          |
          v
   Create SEARCHING Object
          |
          v
    Allocate Array Memory
          |
          v
        Accept()
          |
          v
       Display()
          |
          v
 Initialize iStart and iEnd
          |
          v
 Compare Elements From
       Both Ends
          |
       +--+--+
       |     |
     Match  No Match
       |     |
       v     v
 Return 1  Move iStart
           and iEnd
              |
              v
       iStart <= iEnd ?
          |        |
         Yes       No
          |        |
          +----+---+
               |
               v
           Return 0
               |
               v
         Display Result
               |
               v
             free()
               |
               v
              Stop

===============================================================================
Time Complexity:
===============================================================================

Best Case    : O(1)   → Element found at either end
Average Case : O(n)
Worst Case   : O(n)   → Element found near the middle or absent

Space Complexity : O(n)

Where n is the size of the dynamically allocated array.

===============================================================================
Conclusion:
===============================================================================

Program 004 implements Bi-Directional Search in C using a dynamically
allocated array. The program compares elements from both the beginning
and the end of the array in each iteration. This reduces the number of
iterations compared with checking only one side at a time in some cases,
while the overall worst-case time complexity remains O(n).

===============================================================================
*/

File: Java/Searching/Searching_04.java
/*
===============================================================================
Program Name    : Searching - Linear Search
Program Number  : 04 
File Name       : Searching_04.java

Description     : This program accepts elements from the user, displays the
                  array elements and searches for a particular element using
                  the Linear Search technique.

Date            : August 2026
Language        : Java
Author          : Ishwari Vijaykumar Surve
===============================================================================

Concepts Used:
1. Class
2. Object-Oriented Programming (OOP)
3. Constructor
4. Array
5. Scanner Class
6. User Input
7. Accept Operation
8. Display Operation
9. Linear Search
10. Boolean Return Value
11. Array Traversal

Key Learning Points:
- Understand the basic concept of Linear Search.
- Learn how to search an element sequentially in an array.
- Understand comparison of each array element with the search value.
- Learn how to return true when an element is found.
- Learn how to return false when an element is not found.
- Understand the best and worst cases of Linear Search.

===============================================================================
*/

import java.util.*;

class Searching
{
    private int Arr[];
    private int iSize;

    public Searching(int iNo)
    {
        iSize = iNo;
        Arr = new int[iSize];
    }

    ///////////////////////////////////////////////////////////////////////////
    //
    // Function Name : Accept
    // Description   : Accept elements from user
    //
    ///////////////////////////////////////////////////////////////////////////

    public void Accept()
    {
        Scanner sobj = new Scanner(System.in);

        int i = 0;

        System.out.println("Enter the Elements : ");

        for(i = 0; i < iSize; i++)
        {
            Arr[i] = sobj.nextInt();
        }
    }

    ///////////////////////////////////////////////////////////////////////////
    //
    // Function Name : Display
    // Description   : Display elements of the array
    //
    ///////////////////////////////////////////////////////////////////////////

    public void Display()
    {
        int i = 0;

        System.out.println("\nElements of the Array are : ");

        for(i = 0; i < iSize; i++)
        {
            System.out.println(Arr[i]);
        }
    }

    ///////////////////////////////////////////////////////////////////////////
    //
    // Function Name : LinearSearch
    // Description   : Search element sequentially
    //
    ///////////////////////////////////////////////////////////////////////////

    public boolean LinearSearch(int iNo)
    {
        boolean bFlag = false;

        int i = 0;

        for(i = 0; i < iSize; i++)
        {
            if(Arr[i] == iNo)
            {
                bFlag = true;
                break;
            }
        }

        return bFlag;
    }
}

class Searching_04
{
    public static void main(String A[])
    {
        Searching sobj = new Searching(5);

        sobj.Accept();
        sobj.Display();

        if(sobj.LinearSearch(30) == true)
        {
            System.out.println("\nElement is Present");
        }
        else
        {
            System.out.println("\nThere is no such element");
        }
    }
}

/*
===============================================================================
Expected Output:
===============================================================================

Enter the Elements :
11
21
30
51
101

Elements of the Array are :
11
21
30
51
101

Element is Present

===============================================================================
Memory Representation:
===============================================================================

Searching Object

       sobj
        |
        v
+----------------------+
| Searching Object     |
+----------------------+
| iSize = 5            |
| Arr ----------------------+
+----------------------+     |
                             v
                       +----+----+----+----+-----+
                       | 11 | 21 | 30 | 51 | 101 |
                       +----+----+----+----+-----+
                         0    1    2    3     4
                                  ^
                                  |
                            Search Element
                                  30

===============================================================================
Algorithm:
===============================================================================

1. Start the program.
2. Create a Searching object with size 5.
3. Call Accept() to accept array elements.
4. Call Display() to display the array elements.
5. Call LinearSearch() with the element to be searched.
6. Set bFlag to false.
7. Traverse the array from index 0 to iSize - 1.
8. Compare each array element with the given element.
9. If the element is found:
   - Set bFlag to true.
   - Stop the searching process.
10. Return bFlag.
11. If bFlag is true, display "Element is Present".
12. Otherwise, display "There is no such element".
13. Stop the program.

===============================================================================
Program Flow:
===============================================================================

        Start
          |
          v
   Create Searching Object
          |
          v
      Call Accept()
          |
          v
    Accept Array Elements
          |
          v
     Call Display()
          |
          v
   Call LinearSearch()
          |
          v
    Compare Array Element
          |
          v
      Element Found?
        /       \
      Yes        No
       |          |
       v          v
    Return      Check Next
     true       Element
       |          |
       |          v
       |      End of Array?
       |        /     \
       |      Yes      No
       |       |        |
       |       v        |
       |     Return     |
       |     false <----+
       | 
       v
 Display Result
       |
       v
      Stop

===============================================================================
Time Complexity:
===============================================================================

Best Case    : O(1)
Average Case : O(n)
Worst Case   : O(n)

Where n is the number of elements in the array.

===============================================================================
Space Complexity:
===============================================================================

O(n) for storing n integer elements in the array.

Auxiliary Space : O(1)

===============================================================================
Conclusion:
===============================================================================

Program 04 demonstrates Linear Search in Java. The program sequentially
compares each element of the array with the given search element and returns
true when the element is found. If the element is not present, it returns
false. This program helps understand sequential searching and array traversal.

===============================================================================
*/
